### DISCUSSION TITLE
maya-usd interop status - How do you export and import Maya -> USD -> Maya without breaking shading?

### DISCUSSION THREAD

**Discussion by ColinKennedy** [Autodesk/maya-usd#2353] [General]
Our pipeline needs to be able to reimport USD data geometry and shading as Maya nodes and with materials applied so that they can be fed into rigs and animated. However even with a "cube + single material" case, I can't seem to get import / export to work.

I've tried a variety of set-ups and will summarize the results. The permutations are

- Selected nodes to export / differing inputs
- What shader is assigned
- What was used to export
- What was the final result on-import

**Important**: This set of tests were conducted using the latest Maya and maya-usd. Maya 2023 and maya-usd 0.17.0

| Shader / Misc Notes | Selection                               | Export Process | USD → Maya                                                         |
| ------------------- | --------------------------------------- | -------------- |:------------------------------------------------------------------ |
| Lambert             | Just the cube transform                 | USD Export     | Shader doesn't export. Loses shader bindings on import as a result |
| Lambert             | cube transform + shader                 | USD Export     | Shader doesn't export. Loses shader bindings on import as a result |
| Lambert             | cube transform + shader + shading group | USD Export     | Shader doesn't export. Loses shader bindings on import as a result |

I also tried selecting just the cube's shape, shape + transform, also tried setting the Usd Export **"Plug-In Configuration" to "Arnold" or "None"**. No set of options or selection that I could find that exports the Lambert. Could be my user error but I'm not sure what I'm doing wrong.

| Shader / Misc Notes | Selection                                | Export Process | USD → Maya                                                  |
| ------------------- | ---------------------------------------- | -------------- |:----------------------------------------------------------- |
| Lambert             | Just the cube transform                  | Arnold-USD     | Mesh + shader exports. But neither import correctly back in |
| Lambert             | Cube transform + shader                  | Arnold-USD     | Mesh + shader exports. But neither import correctly back in |
| Lambert             | Cube transform + shader + shading group	 | Arnold-USD     | Mesh + shader exports. But neither import correctly back in |

- One difference between these export processes I've noticed. 
    - Unlike USD Export which you could always reliably bring in back into Maya as a maya-usd stage (not import. The procedural stage), Arnold USD may appear completely empty when brought back as a stage. 
	- The UFE maya-usd procedural allows you to specify a "Prim path" for display but couldn't get that going in practice
- Arnold-USD has a "prefix" export option. You can write "/root" to make a parent Xform Prim to place some of your selected Prims. 
    - But it appears to only apply for DAG nodes (the cube)
	- DG nodes instead are given the prefix but not as a parent Prim but it directly prefixes the Prim name
	    - This appears to not always apply on all DG nodes. Only some

```	
#usda 1.0
(
    endFrame = 0
    startFrame = 0
)

def ArnoldOptions "options"
{
    # ... stuff here ...
}

def "rootdefaultArnoldFilter"
{
    def ArnoldGaussianFilter "gaussian_filter"
    {
	    # ... stuff here ...
    }
}

def "rootdefaultArnoldDriver"
{
    def ArnoldDriverExr "driver_exr_RGBA"
    {
	    # ... stuff here ...
    }
}

def ArnoldColorManagerOcio "rootdefaultColorMgtGlobals"
{
    # ... stuff here ...
}

def Xform "root"
{
    def Xform "pCube1"
    {
        def Mesh "pCubeShape1"
        {
		    # ... stuff here ...
		}
	}
}

def "materials"
{
    def Material "rootlambert2"
    {
        token outputs:arnold:surface.connect = </rootlambert2.outputs:surface>
    }
}


def Shader "rootlambert2"
{
    uniform token info:id = "arnold:lambert"
    float inputs:Kd = 0.8
    color3f inputs:Kd_color = (1, 1, 0)
    token outputs:surface
    string primvars:dcc_name = "lambert2" (
        elementSize = 1
        interpolation = "constant"
    )
}
```

In this case, Arnold-Export did export geometry and shading but in my instances, neither a regular USD Import doesn't bring the shader back in

## Conclusion (So Far)

The best result I could get was to not use any menus and just use `cmds.mayaUsdExport`, directly.

`cmds.mayaUSDExport(["|pCube1"], file=path, parentScope="/root")`

I assume mayaUSDExport is what File > Import | Type: USD Export uses so then that means some export options should be doable but I just don't know which it is.

However, this export process has its own problems.

- The model edge normals are smoothed, so hard edged surfaces no longer "look" right
- the shader is brought back in as usdPreviewSurface (at least for lambert and blinn). Arnold shaders don't come back in at all and the default lambert1 is used

I'm sorry if I'm missing something really simple. I don't know where to look. Could someone help show me how to get (geometry + basic shading, Arnold or non-Arnold shading) USD export -> import interop is working?

**Comment by JGamache-autodesk** [comment]
There are a few issues you need to be aware of when dealing with materials.
- The default material is never exported.
- Perfect roundtripping is impossible to implement using USD Preview Surface materials. Roundtripping will improve with MaterialX materials, and should be best with Arnold materials.

Here something that works for me:
- Create cube
- Assign Lambert, make it red
- Select cube
- File -> Export Selection... : "USD Export", make sure the "MaterialX shading" is enabled in the "Materials" section.
- File -> Import

This should result in a red Lambert on the imported cube. We also support a few utility nodes like file textures, and how they are placed.
The normals will still be problematic, but you can improve that by selecting "None (polygonal mesh)" for "Subdivision method" under the "Geometry" folder. This will save the Maya normals and reload them.
